### PR TITLE
Add fast resnet model

### PR DESCRIPTION
**Pull request by Yancey0623**

### CODE REVIEW

File: fluid/PaddleCV/image_classification/fast_resnet/torchvision_reader.py [hunk 1]
```diff
@@ -0,0 +1,233 @@
+import os
+
+import numpy as np
+import math
+import random
+import torch
+import torch.utils.data
+from torch.utils.data.distributed import DistributedSampler
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/torchvision_reader.py]: can remove torch deps here?

File: fluid/PaddleCV/image_classification/fast_resnet/torchvision_reader.py [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+from tqdm import tqdm
+import time
+import multiprocessing
+
+TRAINER_NUMS = int(os.getenv("PADDLE_TRAINER_NUM", "1"))
+TRAINER_ID = int(os.getenv("PADDLE_TRAINER_ID", "0"))
+epoch = 0
+
+FINISH_EVENT = "FINISH_EVENT"
+#def paddle_data_loader(torch_dataset, indices=None, concurrent=1, queue_size=3072, use_uint8_reader=False):
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/torchvision_reader.py]: need clean up this

File: fluid/PaddleCV/image_classification/fast_resnet/train.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
+# Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/train.py]: Readme for model description?

File: fluid/PaddleCV/image_classification/fast_resnet/train.py [hunk 4]
```diff
@@ -0,0 +1,460 @@
...
+            optimizer = None
+            if is_train:
+                epochs = [(0,7), (7,13), (13, 22), (22, 25), (25, 28)]
+                bs_epoch = [x if is_mp_mode() else x * get_device_num() for x in [224, 224, 96, 96, 50]]
+                lrs = [(1.0, 2.0), (2.0, 0.25), (0.42857142857142855, 0.04285714285714286), (0.04285714285714286, 0.004285714285714286), (0.0022321428571428575, 0.00022321428571428573), 0.00022321428571428573]
+                images_per_worker = args.total_images / get_device_num() if is_mp_mode() else args.total_images
+                optimizer = fluid.optimizer.Momentum(
+                    learning_rate=linear_lr_decay_by_epoch(lrs, epochs, bs_epoch, images_per_worker),
+                    momentum=0.9,
+                    regularization=fluid.regularizer.L2Decay(1e-4))
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/train.py]: fastai implementation do not need to decay on bn parameters?

File: fluid/PaddleCV/image_classification/fast_resnet/train.py [hunk 5]
```diff
@@ -0,0 +1,460 @@
...
+        main_program=train_prog,
+        exec_strategy=strategy,
+        build_strategy=build_strategy,
+        num_trainers=num_trainers,
+        trainer_id=trainer_id)
+
+    test_exe = fluid.ParallelExecutor(
+        True, main_program=test_prog, share_vars_from=train_exe)
+
+    #return train_args, test_args, test_prog, train_exe, test_exe
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/train.py]: clean this line.

File: fluid/PaddleCV/image_classification/fast_resnet/train.py [hunk 6]
```diff
@@ -0,0 +1,460 @@
...
+    return train_args, test_args, test_prog, train_exe, test_exe
+
+# NOTE: only need to benchmark using parallelexe
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/train.py]: this comment is no longer needed

File: fluid/PaddleCV/image_classification/fast_resnet/train.py (hunk 4 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/train.py]: fixed, done.

File: fluid/PaddleCV/image_classification/fast_resnet/train.py (hunk 5 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/train.py]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/train.py (hunk 6 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/train.py]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/README.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
+# PaddlePaddle Fast ResNet
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/README.md]: Fast Imagenet training?

File: fluid/PaddleCV/image_classification/fast_resnet/README.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+PaddlePaddle Fast ResNet can train ImageNet with fewer epochs. We implemented the it according to the blog 
+[Now anyone can train Imagenet in 18 minutes](https://www.fast.ai/2018/08/10/fastai-diu-imagenet/) which published on the [fast.ai] website.
+PaddlePaddle Fast ResNet using the dynmiac batch size, dynamic image size, rectangular images validation and etc... so that the FastResNet can achieve the baseline
+(acc1: 75%, acc5: 93%) by 27 epochs on 8 GPUs.
+
+## Experiment
+
+1. Preparing the training data, resize the images to 160 and 352 by `resize.py`, the prepared data folder is as followed:
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/README.md]: Preparing -> Prepare

by -> using

the prepared data folder is as followed: -> the prepared data folder should look like:

File: fluid/PaddleCV/image_classification/fast_resnet/README.md [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+      |-160
+        |-train
+        `-validation
+      `-352
+        |-train
+        `-validation
+    ```
+1. Install the requirements by `pip install -r requirement.txt`.
+1. Launch the training job: `python train.py --data_dir /data/imagenet`
+1. Learning curve, we launch the training job on V100 GPU card:
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/README.md]: 8 * V100?

File: fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+
+PATH = Path('/data/imagenet2')
+DEST = Path('/data/imagenet2/sz')
+def mkdir(path):
+    if not path.exists():
+        path.mkdir()
+
+mkdir(DEST)
+#szs = (160, 352)
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py]: clean this?

File: fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+
+for sz in szs:
+    ssz=str(sz)
+    mkdir((DEST/ssz))
+    for ds in ('validation','train'): mkdir((DEST/ssz/ds))
+    for ds in ('train',): mkdir((DEST/ssz/ds))
+
+#for ds in ('val','train'): resize_imgs(PATH/ds)
+#for ds in ("validation", "train"):
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py]: clean this?

File: fluid/PaddleCV/image_classification/fast_resnet/README.md (hunk 7 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/README.md]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/README.md (hunk 8 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/README.md]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/README.md (hunk 9 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/README.md]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py (hunk 10 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py]: done.

File: fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py (hunk 11 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/fast_resnet/tools/resize.py]: done.

**Review by typhoonzero — APPROVED** [review]
LGTM for now, should remove torch deps later.

**Comment by Yancey0623** [comment]
@typhoonzero sure will do that in another PR.

**Pull request merged**